src/txn/txn_nsnap.c

```diff
@@ -0,0 +1,372 @@
+
+	if (found == NULL)
+		return (WT_NOTFOUND);
+
+	/* Bump the global ID if we are removing the first entry */
+	if (found == STAILQ_FIRST(&txn_global->nsnaph))
+		txn_global->nsnap_oldest_id = (STAILQ_NEXT(found, q) != NULL) ?
+		    STAILQ_NEXT(found, q)->snap_min : WT_TXN_NONE;
+	STAILQ_REMOVE(&txn_global->nsnaph, found, __wt_named_snapshot, q);
+	__nsnap_destroy(session, found);
+
+	return (ret);
+}
+
+/*
+ * __nsnap_drop_to --
+ *	Drop named snapshots, if the name is NULL all snapshots will be
+ *	dropped. The named snapshot lock must be held write locked.
+ */
+static int
+__nsnap_drop_to(WT_SESSION_IMPL *session, WT_CONFIG_ITEM *name, int inclusive)
+{
+	WT_DECL_RET;
+	WT_NAMED_SNAPSHOT *last, *nsnap, *prev;
+	WT_TXN_GLOBAL *txn_global;
+	uint64_t new_nsnap_oldest;
+
+	last = nsnap = prev = NULL;
+	txn_global = &S2C(session)->txn_global;
+
+	if (STAILQ_EMPTY(&txn_global->nsnaph)) {
+		if (name == NULL)
+			return (0);
+		/*
+		 * Dropping specific snapshots when there aren't any it's an
+		 * error.
+		 */
+		WT_RET_MSG(session, EINVAL,
+		    "Named snapshot '%.*s' for drop not found",
+		    (int)name->len, name->str);
+	}
+
+	/*
+	 * The new ID will be none if we are removing all named snapshots
+	 * which is the default behavior of this loop.
+	 */
+	new_nsnap_oldest = WT_TXN_NONE;
+	if (name != NULL) {
+		STAILQ_FOREACH(last, &txn_global->nsnaph, q) {
+			if (WT_STRING_MATCH(last->name, name->str, name->len))
+				break;
+			prev = last;
+		}
+		if (last == NULL)
+			WT_RET_MSG(session, EINVAL,
+			    "Named snapshot '%.*s' for drop not found",
+			    (int)name->len, name->str);
+
+		if (!inclusive) {
+			/* We are done if a drop upto points to the head */
+			if (prev == 0)
+				return (0);
+			last = prev;
+		}
+
+		if (STAILQ_NEXT(last, q) != NULL)
+			new_nsnap_oldest = STAILQ_NEXT(last, q)->snap_min;
+	}
+
+	do {
+		nsnap = STAILQ_FIRST(&txn_global->nsnaph);
+		WT_ASSERT(session, nsnap != NULL);
+		STAILQ_REMOVE_HEAD(&txn_global->nsnaph, q);
+		__nsnap_destroy(session, nsnap);
+	/* Last will be NULL in the all case so it will never match */
+	} while (nsnap != last && !STAILQ_EMPTY(&txn_global->nsnaph));
+
+	/* Now that the queue of named snapshots is updated, update the ID */
+	txn_global->nsnap_oldest_id = new_nsnap_oldest;
+
+	return (ret);
+}
+
+/*
+ * __wt_txn_named_snapshot_begin --
+ *	Begin an named in-memory snapshot.
+ */
+int
+__wt_txn_named_snapshot_begin(WT_SESSION_IMPL *session, const char *cfg[])
+{
+	WT_CONFIG_ITEM cval;
+	WT_DECL_RET;
+	WT_NAMED_SNAPSHOT *nsnap, *nsnap_new;
+	WT_TXN *txn;
+	WT_TXN_GLOBAL *txn_global;
+	const char *txn_cfg[] =
+	    { WT_CONFIG_BASE(session, WT_SESSION_begin_transaction),
+	      "isolation=snapshot", NULL };
+	int started_txn;
+
+	started_txn = 0;
+	nsnap_new = NULL;
+	txn_global = &S2C(session)->txn_global;
+	txn = &session->txn;
+
+	WT_RET(__wt_config_gets_def(session, cfg, "name", 0, &cval));
+	WT_ASSERT(session, cval.len != 0);
+
+	if (!F_ISSET(txn, WT_TXN_RUNNING)) {
```

Comment:
The transaction begin code/logic is currently within the nsnap write lock. It doesn't need to be here, but it was simpler to leave it inline with the rest of snapshot begin (if we split this out we'd need a different way to determine whether we started the transaction).